src/lib/get-agent.ts

```diff
@@ -0,0 +1,75 @@
+import CONFIG from "../config";
+
+const { USER_AGENT, MCP_USER_AGENT } = CONFIG.CLIENT_SETTINGS;
+
+/**
+ * Checks if the main module filename indicates MCP context.
+ * @returns true if main module contains "mailtrap-mcp" and is not in node_modules
+ */
+function isMainModuleMCP(): boolean {
+  const mainFile = require?.main?.filename;
+
+  return !!(
+    mainFile &&
+    mainFile.includes("mailtrap-mcp") &&
+    !mainFile.includes("node_modules")
+  );
+}
+
+/**
+ * Checks if running in MCP runtime context (Claude Desktop).
+ * @returns true if main module is from MCP runtime
+ */
+function isMCPRuntimeContext(): boolean {
+  const mainFile = require?.main?.filename;
+
+  return !!(
+    mainFile &&
+    (mainFile.includes("mcp-runtime") ||
+      mainFile.includes("nodeHost.js") ||
+      mainFile.includes("Claude.app"))
+  );
+}
+
+/**
+ * Checks if the current working directory indicates MCP context.
+ * @returns true if cwd contains "mailtrap-mcp" and is not in node_modules
+ */
+function isWorkingDirectoryMCP(): boolean {
+  try {
+    const cwd = process.cwd();
+    return cwd.includes("mailtrap-mcp") && !cwd.includes("node_modules");
+  } catch {
+    return false;
+  }
+}
+
+/**
+ * Checks if the call stack indicates MCP context.
+ * @returns true if stack contains "mailtrap-mcp" and is not from node_modules/mailtrap
+ */
+function isCallStackMCP(): boolean {
+  const { stack } = new Error();
+
+  return !!(
+    stack &&
+    stack.includes("mailtrap-mcp") &&
+    !stack.includes("node_modules/mailtrap")
+  );
+}
+
+/**
+ * Gets the appropriate User-Agent string based on the current context.
+ * @returns The User-Agent string for the current context
+ */
+function getDynamicUserAgent(): string {
```

Comment:
If you ask me, bringing the MCP concern to the SDK project is a violation of the SRP 🙂 I would approach this differently and allow specifying the user agent directly via a configuration setting of the SDK client. And, override this setting in the MCP server NPM package.
It's simpler, and it's not a big deal to allow users of the SDK to override the user agent; it's fairly easy to do this anyway.
WDYT?

Reply:
if allowing users to override user agent is not a problem, then its simple solution. I thought that we don't want to enable users to change the user agent.

Reply:
I agree with @vittorius, it's cleaner to allow to override user agent, and it might be useful for other purposes (for the users), too.